crypto/provider_conf.c

```diff
@@ -313,3 +313,49 @@ void ossl_provider_add_conf_module(void)
     OSSL_TRACE(CONF, "Adding config module 'providers'\n");
     CONF_module_add("providers", provider_conf_init, NULL);
 }
+
+static CONF *ossl_conf_from_params(OSSL_LIB_CTX *ctx, OSSL_PARAM *params)
+{
+    CONF *my_conf = NCONF_new_ex(ctx, NCONF_default());
+    BIO *conf_bio = BIO_new(BIO_s_mem());
+    int ok = 0;
+    int i;
+
+    if (BIO_printf(conf_bio, "[%s]\nactivate = 1\n", "my_sect") < 0)
+        goto err;
+
+    for (i = 0; params[i].key != NULL; i++) {
+        if (params[i].data_type != OSSL_PARAM_UTF8_STRING)
+            continue;
+        if (BIO_printf(conf_bio, "%s = %s\n", params[i].key, (char *)params[i].data) <= 0)
+            goto err;
+    }
+
+    if (NCONF_load_bio(my_conf, conf_bio, NULL) != 1)
+        goto err;
+
+    ok = 1;
+err:
+    if (ok != 1) {
+        NCONF_free(my_conf);
+        my_conf = NULL;
+    }
+    BIO_free(conf_bio);
+
+    return my_conf;
+}
+
+int OSSL_PROVIDER_activate(OSSL_LIB_CTX *libctx, const char *name,
+                                OSSL_PARAM *params)
+{
+    CONF *my_conf = ossl_conf_from_params(libctx, params);
```

Comment:
I find this implementation quite surprising. It is a way to create a config file from OSSL_PARAMs - but I understood the requirement to be to be able to configure providers from params. This takes params and converts them into a config file. Later that will be converted back into OSSL_PARAMs to pass to the provider. Why not just pass the OSSL_PARAMs straight to the provider without the config file middle-man.

Reply:
Yes. It's a quite ugly and hackerish (in bad sense) way to reach my goal.

I didn't find an internal API function passing params to provider -  `OSSL_PROVIDER_load` doesn't accept params and functions from `crypto/provider_conf.c` deal with parsed config, not params.

Reply:
Instead of creating conf, you could create the sk_INFOPAIR stack of name/value items that the provider could query via the core_get_params() call. Or you could even modify the core_get_params() to return params directly but that would be more invasive change.

Reply:
What about adding get/set params calls at the provider level?

If we're using OSSL_PARAM as the passed type (& I believe we should since it was designed for this), let's just use it properly without any kludges.

Reply:
That would require changing providers. IMO not worth it.

And we already do get params but as a callback from provider to core.

This proposal already does the OSSL_PROVIDER_load_ex() albeit with strange OSSL_PROVIDER_activate() name.

Reply:
`OSSL_PROVIDER_load` returns a provider handle and it would be strange if `OSSL_PROVIDER_load_ex` returns anything different. But I agree that the proposed name is bad.

Reply:
PARAMS were designed for the cross provider interface.  Using anything else means reinventing this or bodging something else in it's place.

We already use params to configure providers -- look at the options for the FIPS provider.

IMO, this is the only sensible way forwards.

Reply:
I agree but how do we pass them otherwise then in config form?

Reply:
I'm not sure what you mean here.
How to convert config to params keeping the nesting?
Or something else?

Reply:
We have an internal function (I extracted it myself :) ) that accepts a config structure and activates the provider. It's not clear to me how to pass the parameters directly to provider so that's why I dance this strange dance to use the maximum of the current code.

Reply:
We will need to discuss this offline I suspect ...

Reply:
Resolving this thread because the function is implemented in a normal way now.